// fsh-sushi/dist/fshtypes/FshCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FshEntity_1 = require("./FshEntity");
class FshCode extends FshEntity_1.FshEntity {
    constructor(code, system, display) {
        super();
        this.code = code;
        this.system = system;
        this.display = display;
    }
    toString() {
        var _a;
        const str = `${_a = this.system, (_a !== null && _a !== void 0 ? _a : '')}#${this.code}`;
        return this.display ? `${str} "${this.display}"` : str;
    }
    toFHIRCoding() {
        const coding = {};
        if (this.code) {
            coding.code = this.code;
        }
        if (this.system) {
            if (this.system.indexOf('|') > -1) {
                [coding.system, coding.version] = this.system.split('|', 2);
            }
            else {
                coding.system = this.system;
            }
        }
        if (this.display) {
            coding.display = this.display;
        }
        return coding;
    }
    toFHIRCodeableConcept() {
        return {
            coding: [this.toFHIRCoding()]
        };
    }
    toFHIRQuantity() {
        const quantity = {};
        if (this.code) {
            quantity.code = this.code;
        }
        if (this.system) {
            quantity.system = this.system;
        }
        if (this.display) {
            quantity.unit = this.display;
        }
        return quantity;
    }
}
exports.FshCode = FshCode;
