// fsh-sushi/dist/fshtypes/FshEntity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Abstract class to contain elements common to all FSH types.
 */
class FshEntity {
    constructor() {
        this.sourceInfo = {};
    }
    withLocation(location) {
        if (Array.isArray(location)) {
            this.sourceInfo.location = {
                startLine: location[0],
                startColumn: location[1],
                endLine: location[2],
                endColumn: location[3]
            };
        }
        else {
            this.sourceInfo.location = location;
        }
        return this;
    }
    withFile(file) {
        this.sourceInfo.file = file;
        return this;
    }
    withAppliedLocation(location) {
        if (Array.isArray(location)) {
            this.sourceInfo.appliedLocation = {
                startLine: location[0],
                startColumn: location[1],
                endLine: location[2],
                endColumn: location[3]
            };
        }
        else {
            this.sourceInfo.appliedLocation = location;
        }
        return this;
    }
    withAppliedFile(file) {
        this.sourceInfo.appliedFile = file;
        return this;
    }
}
exports.FshEntity = FshEntity;
