// fsh-sushi/dist/fshtypes/ValueSetComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FshEntity_1 = require("./FshEntity");
class ValueSetComponent extends FshEntity_1.FshEntity {
    constructor(inclusion) {
        super();
        this.inclusion = inclusion;
        this.from = {};
    }
}
exports.ValueSetComponent = ValueSetComponent;
class ValueSetConceptComponent extends ValueSetComponent {
    constructor() {
        super(...arguments);
        this.concepts = [];
    }
}
exports.ValueSetConceptComponent = ValueSetConceptComponent;
class ValueSetFilterComponent extends ValueSetComponent {
    constructor() {
        super(...arguments);
        this.filters = [];
    }
}
exports.ValueSetFilterComponent = ValueSetFilterComponent;
var VsOperator;
(function (VsOperator) {
    VsOperator["EQUALS"] = "=";
    VsOperator["IS_A"] = "is-a";
    VsOperator["DESCENDENT_OF"] = "descendent-of";
    VsOperator["IS_NOT_A"] = "is-not-a";
    VsOperator["REGEX"] = "regex";
    VsOperator["IN"] = "in";
    VsOperator["NOT_IN"] = "not-in";
    VsOperator["GENERALIZES"] = "generalizes";
    VsOperator["EXISTS"] = "exists";
})(VsOperator = exports.VsOperator || (exports.VsOperator = {}));
