// fsh-sushi/dist/export/FHIRExporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
/**
 * FHIRExporter handles the processing of FSH documents, storing the FSH types within them as FHIR types.
 * FHIRExporter takes the Profiles and Extensions within the FSHDocuments of a FSHTank and returns them
 * as a structured Package.
 */
class FHIRExporter {
    constructor(tank, pkg, fisher) {
        this.tank = tank;
        this.pkg = pkg;
        this.fisher = fisher;
        this.structureDefinitionExporter = new _1.StructureDefinitionExporter(this.tank, this.pkg, this.fisher);
        this.valueSetExporter = new _1.ValueSetExporter(this.tank, this.pkg, this.fisher);
        this.codeSystemExporter = new _1.CodeSystemExporter(this.tank, this.pkg, this.fisher);
        this.instanceExporter = new _1.InstanceExporter(this.tank, this.pkg, this.fisher);
        this.mappingExporter = new _1.MappingExporter(this.tank, this.pkg, this.fisher);
    }
    export() {
        this.structureDefinitionExporter.export();
        this.valueSetExporter.export();
        this.codeSystemExporter.export();
        this.instanceExporter.export();
        this.mappingExporter.export();
        return this.pkg;
    }
}
exports.FHIRExporter = FHIRExporter;
