// fsh-sushi/dist/utils/FSHLogger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const winston_1 = require("winston");
const chalk_1 = __importDefault(require("chalk"));
const { combine, printf } = winston_1.format;
const withLocation = winston_1.format(info => {
    if (info.file) {
        info.message += `\n  File: ${info.file}`;
        delete info.file;
    }
    if (info.location) {
        info.message += `\n  Line: ${info.location.startLine}`;
        if (info.location.endLine !== info.location.startLine) {
            info.message += ` - ${info.location.endLine}`;
        }
        delete info.location;
    }
    if (info.appliedFile) {
        info.message += `\n  Applied in File: ${info.appliedFile}`;
        delete info.appliedFile;
    }
    if (info.appliedLocation) {
        info.message += `\n  Applied on Line: ${info.appliedLocation.startLine}`;
        if (info.appliedLocation.endLine !== info.appliedLocation.startLine) {
            info.message += ` - ${info.appliedLocation.endLine}`;
        }
        delete info.appliedLocation;
    }
    return info;
});
const incrementCounts = winston_1.format(info => {
    switch (info.level) {
        case 'info':
            exports.stats.numInfo++;
            break;
        case 'warn':
            exports.stats.numWarn++;
            break;
        case 'error':
            exports.stats.numError++;
            break;
        case 'debug':
            exports.stats.numDebug++;
            break;
        default:
            break;
    }
    return info;
});
const printer = printf(info => {
    let level;
    switch (info.level) {
        case 'info':
            level = chalk_1.default.whiteBright.bgGreen(`${info.level} `);
            break;
        case 'warn':
            // (179, 98, 0) = dark dark orange
            level = chalk_1.default.whiteBright.bgRgb(179, 98, 0)(`${info.level} `);
            break;
        case 'error':
            level = chalk_1.default.whiteBright.bgRed(`${info.level}`);
            break;
        case 'debug':
            level = chalk_1.default.whiteBright.bgBlue(`${info.level}`);
            break;
        default:
            break;
    }
    return `${level} ${info.message}`;
});
exports.logger = winston_1.createLogger({
    format: combine(incrementCounts(), withLocation(), printer),
    transports: [new winston_1.transports.Console()]
});
class LoggerStats {
    constructor() {
        this.numInfo = 0;
        this.numWarn = 0;
        this.numError = 0;
        this.numDebug = 0;
    }
    reset() {
        this.numInfo = 0;
        this.numWarn = 0;
        this.numError = 0;
        this.numDebug = 0;
    }
}
exports.stats = new LoggerStats();
